Redo CLI flags using postgres connection URL.

mberhault opened the pull request:
Fixes #3748.

This changes the CLI args for commands that need SQL connection
parameters (sql, user, zone).

Two ways to specify parameters:
- fully-specified URL (all other flags are ignored)
- individual flags (user, host, port, dbname, insecure, certs)

The default is empty URL and flags corresponding to:
postgresql://root@localhost:15432/?sslmode=verify-all&<all three
cert/key paths>

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3757)

<!-- Reviewable:end -->

mberhault commented on cli/flags.go:
```diff
@@ -61,6 +65,10 @@ var flagUsage = map[string]string{
 
           --attrs=us-west-1b,gpu.
 `,
+	"balance-mode": `
```
most of those are just being moved around to keep them in alphabetical order.

tamird commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mop_tMXl2yP9YGhyY:-K7mog3XyW-2W8iplqdo:702229481) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L32)):*
So we're doing this so we can expose `--port` and `--host` to look more like existing tools? I think maybe I'd prefer `--pgaddr` for symmetry with the server.

---

*[cli/flags.go, line 68 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7miTGYFRCzMQ6GqDk0-r1-68:-K7mo0-V4edfWmB3nKlQ:-477585119) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L68)):*
nit: put this change in a separate commit to ease reviews

---

*[cli/sql_util.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mp7TiT2Q0UAidNwxU:-K7mp7TiT2Q0UAidNwxV:478351269) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/sql_util.go#L38)):*
to some extent, this is duplicating logic present in `pgUrl` in the `sql` package. can we reduce this duplication?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mop_tMXl2yP9YGhyY:-K7qlyEN-u_53EVuzlCY:226777602) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L32)):*
I actually think we should remove the `--addr` and `--pgaddr` flags. No one does it this way. We should have just plain `--port` and `--pgport`, and if we really want, an interface or address to listen on. Combining both is just terrible.

---

*[cli/flags.go, line 68 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7miTGYFRCzMQ6GqDk0-r1-68:-K7qmAV5o9xCpIjaSGY3:1020839866) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L68)):*
will do.

---

*[cli/sql_util.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mp7TiT2Q0UAidNwxU:-K7qmBtV70GAs_cOQJ5F:-901868625) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/sql_util.go#L38)):*
maybe, although if it allows me to avoid a dependency on cockroach/sql, it may be best to leave it this way

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[cli/flags.go, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mop_tMXl2yP9YGhyY:-K7qmi6yOyn71Nq1gQOg:1054307355) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L32)):*
OK, maybe, but that's not what's currently being proposed here.

---

*[cli/sql_util.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mp7TiT2Q0UAidNwxU:-K7qmt4dSUb9H3UizNS9:1032269455) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/sql_util.go#L38)):*
FYI @mrtracy extracted that function from `sql` in https://github.com/cockroachdb/cockroach/pull/3687

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mop_tMXl2yP9YGhyY:-K7r1P1xCNAyJSHSNJhr:1228250873) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L32)):*
this PR drops `--addr` for the client-side sql commands in favor of `--host` and `--port`. `--pgaddr` was already server-side only, so no change there so far. What I would propose (probably in a separate PR as it will involve some plumbing) is changing this for server-side as well.

---

*[cli/sql_util.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mp7TiT2Q0UAidNwxU:-K7r1x1O8tnUhArC9O2B:614554971) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/sql_util.go#L38)):*
ah. I'll take a look.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[cli/flags.go, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mop_tMXl2yP9YGhyY:-K7r2FHOiJGskAR_BJ6j:-2008116144) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L32)):*
> this PR drops `--addr` for the client-side sql commands in favor of `--host` and `--port`.

I see. I still disagree with that; I think we should maintain server/client symmetry in these flags whenever possible.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

Commit:
Rework cli flags.

* remove `--addr` and `--pgaddr`
* introduce `host`, `port`, `pgport` for clients and servers
* support postgres URL, or individual flags on `sql`
* use `pq` driver instead of `postgres` driver
* rework credentials to allow `cli_test` to work with `pq`

mberhault commented:
ok. addressed comments (with caveats):
* split commits (that was kindda tedious after the fact)
* looking into dropping `--addr` for the server side. There's a lot more plumbing and verification there. I'll file an issue against myself and do it.
* Matt's change isn't in. It may or may not be worth sharing that code.

tamird commented:
OK, can you file an issue to share the code once matt's change is in? LGTM otherwise

mberhault commented:
ugh! this is too gross. I'll go ahead and split host/port now.

I'm now proposing the following three flags:
server-side:
* host: address to listen on
* port: port to server postgres on
* cockroach-port: port for cockroach stuff: gossip, KV, and UI

client-side sql commands (eg: sql, user, zone):
* host: address to connect to
* port: postgres port to connect to

client-side KV/admin commands (eg: kv, range, exterminate, quit)
* host: address to connect to
* cockroach-port: cockroach port to connect to

mberhault commented:
had to disable the `Example_zone` and `Example_user` examples in `cli/cli_test.go` since we're now using pgwire but don't have certificate files, just the embedded certs. I'll either copy them to a file (`sql/pgwire_test` does this) or run in insecure mode.

tamird commented:
FYI @mrtracy's PR landed, so you should be able to use pgUrl now

mberhault commented:
ok. getting there. the zone and user examples are almost ready to re-enable, except for the fact that things crash. this seems related to the change to pgwire. I'll investigate more and ping back here.

I haven't merged PGUrl as it does too much custom stuff.

tamird commented on security/securitytest/securitytest.go:
```diff
@@ -17,25 +17,30 @@
-// temporary file.
+// RestrictedCopy creates an on-disk copy of the embedded security asset
+// with the provided path. The copy will be created in the provided directory.
+// Returns the path of the file and a cleanup function that will delete the file.
```
should probably change the word "file" here and on the next line.

mberhault replied:
I'm not sure what you mean. I renamed from "temporary file" to "file". It is still a file, so I'm not sure what I should change it to

tamird commented on sql/logic_test.go:
```diff
@@ -231,7 +237,7 @@ func (t *logicTest) run(path string) {
 
 	// Make a temporary directory to hold all our on-disk crypto assets.
-	tempDir, err := ioutil.TempDir(os.TempDir(), "TestLogic")
+	tempDir, err := ioutil.TempDir("", "TestLogic")
```
i know it's equivalent, but why?

mberhault replied:
because we do that almost everywhere: (from head)
```
git grep 'ioutil.TempDir('
acceptance/cluster/localcluster.go:     l.CertsDir, err = ioutil.TempDir(pwd, ".localcluster.certs.")
acceptance/cluster/localcluster.go:             l.logDir, err = ioutil.TempDir(pwd, ".localcluster.logs.")
acceptance/main_test.go:                logDir, err = ioutil.TempDir(os.TempDir(), "clustertest_")
gossip/simulation/gossip.go:    dirName, err := ioutil.TempDir("", "gossip-simulation-")
server/status_test.go:  dir, err := ioutil.TempDir("", "local_log_test")
sql/logic_test.go:      tempDir, err := ioutil.TempDir(os.TempDir(), "TestLogic")
util/log/main_test.go:  tmpDir, err := ioutil.TempDir("", "logtest")
util/testing.go:        dir, err := ioutil.TempDir("", prefix)
util/testing.go:                dirs[i], err = ioutil.TempDir("", prefix)
```

mberhault commented on cli/sql_util.go:
```diff
@@ -58,13 +71,20 @@ func runQuery(db *sql.DB, query string, parameters ...interface{}) (
 func runQueryWithFormat(db *sql.DB, format fmtMap, query string, parameters ...interface{}) (
 	[]string, [][]string, error) {
-	rows, err := db.Query(query, parameters...)
+	if insertPattern.MatchString(query) {
```
I want to re-do the "pretty" querying. ideally, statements should display the number of rows affected.
Leaving this refactor for later though.

nvanbenschoten commented:
Would it make sense to migrate over [this usage](https://github.com/cockroachdb/cockroach/blob/master/sql/main_test.go#L99) of the cockroach sql/driver to the postgres sql/driver in this PR, or a subsequent one?

mberhault commented:
that one still works, so I'd say a separate one. This was more about fixing the interactive shell.

mberhault commented:
ping

tamird commented:
Reviewed 1 of 6 files at r3, 1 of 4 files at r4, 4 of 5 files at r5, 1 of 5 files at r6, 2 of 3 files at r7, 7 of 7 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 81 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRAFaQeFKMzBx1r7W:-K8tRAFbmIEKtfN5LiOb:-1405618511) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/cli_test.go#L81)):*
"i" is a weird name for this binding

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K8tRW3sw7oF19Y_bMRx:-1956915720) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
not integers?

---

*[cli/flags.go, line 302 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRefnifuJAiYSg-qG:-K8tRefnifuJAiYSg-qH:1918538038) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L302)):*
this is fragile, you should use `net.JoinHostPort`

---

*[cli/sql.go, line 67 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRrBWZd3xDjG1emg7:-K8tRrBXm8DcST1pIvvl:-347592487) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/sql.go#L67)):*
missing verb. "is"?

---

*[sql/logic_test.go, line 240 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7wcKDfTyswcSaj6cyz-r7-240:-K8tSVuSBOzmkr0CX2TK:-1861998450) ([raw file](https://github.com/cockroachdb/cockroach/blob/f43b5a35f96c09adbb471854c5fcf523e53139fa/sql/logic_test.go#L240)):*
I see. Seems like `acceptance/main_test.go` is the last remaining case that doesn't follow this. Change that too?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K8tWMKdcO6jjP66r9qS:-1074550443) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
I'm not fond of naming this `cockroach-port`? I'd prefer to see this flag named `port` and the existing `port` flag renamed `pgport`. 

---

*[cli/flags.go, line 302 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRefnifuJAiYSg-qG:-K8tWbV5EF6jwc2xKuph:-2107861138) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L302)):*
Fragile? The host port separator is unlikely to change, though I agree that using `net.JoinHostPort` would be preferable. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 302 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRefnifuJAiYSg-qG:-K8tZ-ITPBgoh34Zovdk:1222754815) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L302)):*
There's some weird case with ipv6 that I recently came across. I don't recall the details now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 107 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tXXaMIa8Ow_Vpg3fz:-K8tXXaNHa3NtJumnk99:-1489806349) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/cli_test.go#L107)):*
Do you really want to continue after these errors?

---

*[cli/cli_test.go, line 118 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tXk0t9qeBPxw2G-8D:-K8tXk0t9qeBPxw2G-8E:-631143547) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/cli_test.go#L118)):*
Can you always pass both `--cockroach-port` and `--port` instead of switching based on which command is used?

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K8tY92TayHR6rpwYVSj:-740138631) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
`+1`. Or `--cockroach-port` could be `--http-port` since that's the user-visible part of that port's usage.  But since we intend to multiplex as much as possible over that port we might as well call it `--port`. 

---

*[cli/flags.go, line 302 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRefnifuJAiYSg-qG:-K8tZ-faHU4NIo7bv0zY:-1895568180) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L302)):*
`net.JoinHostPort` handles ipv6 literals correctly; this does not.

We should probably split the `context.Addr` field so we don't have to mash the structured flags back into a single string only to parse them out later.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Reviewed 1 of 3 files at r1.
Review status: 7 of 16 files reviewed at latest revision, 10 unresolved discussions.

---

*[cli/cli_test.go, line 81 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRAFaQeFKMzBx1r7W:-K92yndF0cfLgI010shg:2063283053) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/cli_test.go#L81)):*
changed to a

---

*[cli/cli_test.go, line 107 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tXXaMIa8Ow_Vpg3fz:-K92zAikGRpJylqXAQkO:644503926) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/cli_test.go#L107)):*
not sure. I don't particularly want to fail either. I was going with the existing logic.

---

*[cli/cli_test.go, line 118 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tXk0t9qeBPxw2G-8D:-K92z2I-zXE43zu-tJ06:1118653795) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/cli_test.go#L118)):*
not all commands take both, intentionally. We could make them global flags, but I like having the help message list exactly the flags that matter.

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K92zXoYdfWWz7xE19La:2028012487) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
I went for `--port` for the postgres one since that's what most  people will be interacting with. `--http-port` is weird (it may be https or rpc).

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K93-0AJjzaLcVyCq2gb:1175465155) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
they definitely have to be now for `net.JoinHostPort`. I could `itoa`, but that's not any better

---

*[cli/flags.go, line 302 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRefnifuJAiYSg-qG:-K92zyjNu0AWrRrxyWR3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L302)):*
Done.

---

*[cli/sql.go, line 67 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRrBWZd3xDjG1emg7:-K93-FW5cHPtCOz06gak:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/sql.go#L67)):*
Done.

---

*[sql/logic_test.go, line 240 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7wcKDfTyswcSaj6cyz-r7-240:-K93-SErdP3S8HeCAbjk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f43b5a35f96c09adbb471854c5fcf523e53139fa/sql/logic_test.go#L240)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 9 of 9 files at r9, 5 of 5 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K93PdeGdbfumSgg__Vz:913585109) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
Well, it pushes the integer validation to the flags package, rather than kicking the can to some method in "net".

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K93XxK22eDZRckobk08:1523391804) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
I don't think that's necessary. Also, it's nice to allow port names (eg: http instead of 80). Especially once the official cockroach port gets propagated.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K93_T7F38My1EO6xuXT:-175083555) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
>  Also, it's nice to allow port names (eg: http instead of 80)

is that a thing? Anyway, I don't feel strongly enough to keep going back and forth.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 16 of 17 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K93d3lo7Okex58gxd8-:-1398476316) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
yup. it's kindda neat. I can add the following line to `/etc/services`:
```
cockroach       26257/tcp
```

And now run the cmdline as follows (with a `cockroach start --dev` on the side):
```
$ ./cockroach range --dev ls --cockroach-port=cockroach
/Min-/Table/11 [1]
	0: node-id=1 store-id=1
etc...
```

And magic!

of course, without the entry in the `services` file, I get a plain old:
```
$ ./cockroach range --dev ls --cockroach-port cockroach
Error: failed to initialize KV client: unknown port tcp/cockroach
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 203 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tRW3sw7oF19Y_bMRw:-K93glnvbkU4XF859xTI:1667390607) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L203)):*
that error is actually pretty friendly. OK, I'm +1 on this being a string.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 211 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K93jYHPuU02Z7Cf5bL4:-K93jYHQ76OcQGoSMYYf:-957817726) ([raw file](https://github.com/cockroachdb/cockroach/blob/332dd2300f7b764b71922a334476cddd2289b2fc/cli/cli_test.go#L211)):*
I thought Andrei recently got rid of all the explicit `"quit"` commands. Why reintroduce them?

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K93ifgqmEFP8jaJna7R:-1991163385) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
And when we add support for the mysql wire protocol, or the the spanner wire protocol, what should we call those ports.

I'm very against naming this flag `cockroach-port`. I still prefer `port` and find `http-port` acceptable. 

---

*[cli/flags.go, line 138 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K93jHRKk8HtmQPznfJn:-K93jHRKk8HtmQPznfJo:-731004384) ([raw file](https://github.com/cockroachdb/cockroach/blob/332dd2300f7b764b71922a334476cddd2289b2fc/cli/flags.go#L138)):*
The help text even identifies this as the port for `Postgres` traffic. `pgport`!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 211 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K93jYHPuU02Z7Cf5bL4:-K93kN2kK84Exwze8bBr:1737939334) ([raw file](https://github.com/cockroachdb/cockroach/blob/332dd2300f7b764b71922a334476cddd2289b2fc/cli/cli_test.go#L211)):*
because I would need two stops: c.stop() to cleanup the copies of the test resources, and c.Stop() to stop the testserver. but I can't always call both (eg: in `Example_quit`). I may have to change it back to Andrei's version as I'm hitting races again.

---

*[cli/flags.go, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K7mop_tMXl2yP9YGhyY:-K93kljVoUynb9DBoNp7:-1027858871) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8ff2034d95ba3bd9ec4ed413ef6bd72f383865c/cli/flags.go#L32)):*
I think this is addressed. I just never marked it as done.

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K93kp6XGjSVJDo6DA_n:-259126280) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
Ok, let's summarize the options:
1) this PR: `port` for postgres, `cockroach-port` for cockroach
2) `port` for postgres and `http-port` for cockroach (I find the latter misleading)
3) `pgport` for postgres and `port` for cockroach (I really really prefer `port` for postgres, this is what people will be using primarily)

---

*[cli/flags.go, line 138 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K93jHRKk8HtmQPznfJn:-K93kupKQH2k9oSQb3c1:1514963494) ([raw file](https://github.com/cockroachdb/cockroach/blob/332dd2300f7b764b71922a334476cddd2289b2fc/cli/flags.go#L138)):*
(addressing in other comment)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K93mBZBw8LBspcQErGj:1416942209) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
Understood. My preference is 3, 2, 1. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K93myB6n1OxLW5ab_MP:-57625007) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
Agree with 3,2,1.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 16 of 17 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 211 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K93jYHPuU02Z7Cf5bL4:-K93twPujLDoGG4E6jh6:805131656) ([raw file](https://github.com/cockroachdb/cockroach/blob/332dd2300f7b764b71922a334476cddd2289b2fc/cli/cli_test.go#L211)):*
ok. reverted back and changed to a manual cleanup in TestQuit.

---

*[cli/flags.go, line 78 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K8tWMKdcO6jjP66r9qR:-K93u-pxVeyLVnIzRtTb:-726041319) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8138d9e52e4a74d01fbf621c8b45402eab10aad/cli/flags.go#L78)):*
right then. I'll change to `pgport` and `port`. Will ping when done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
This is done. lots and lots of merge conflicts (mostly on flag.go). May have to tweak some more.

tamird commented:
Reviewed 3 of 4 files at r13, 4 of 4 files at r14.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[cli/cli_test.go, line 540 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98B3qoTPsp1vRnwWYb:-K98B3qoTPsp1vRnwWYc:-797296808) ([raw file](https://github.com/cockroachdb/cockroach/blob/393c4a8e32f3d443441d0e68c0d903be3f813536/cli/cli_test.go#L540)):*
looks like a mistake?

---

*[cli/flags.go, line 25 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98AfGbm00FcuMx52Hs:-K98AfGccI0SWHQ993ae:485074348) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a5ef98ef4a86afdf7539e6b9adbb40d54e9d3c/cli/flags.go#L25)):*
why'd this move?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented on cli/cli_test.go:
```diff
@@ -449,8 +515,6 @@ range_max_bytes: 67108864
 	// - attrs: [us-west-1b, ssd]
 	// range_min_bytes: 8388608
 	// range_max_bytes: 67108864
-	// gc:
```
what happened here?

mberhault replied:
oh. I added it to see if I could change the bytes representation of the zone config. no dice. I commented it out instead (another bug to file)

tamird replied:
I'm confused. Looks like you commented out the whole test (which is fine) but this change to the test body looks like a rebase error.

mberhault replied:
oh. sorry, right you are. What I had originally modified was the `zone100 = ...` line, not the example output. Reverted.

tamird commented on cli/flags.go:
```diff
@@ -267,6 +279,24 @@ func initFlags(ctx *Context) {
 		f.BoolVarP(&ctx.OneShotSQL, "execute", "e", ctx.OneShotSQL, flagUsage["execute"])
 	}
 
+	// Commands that need the cockroach port.
```
i'm confused, isn't this exactly `clientCmds`?

tamird replied:
oh, disregard!

tamird commented:
LGTM modulo some rebase shenanigans

mberhault commented:
thanks. Broken tests, and rebase cleanup. There's enough shuffling of things back and forth that I may collapse all.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cli/cli_test.go, line 540 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98B3qoTPsp1vRnwWYb:-K98LueoQ8pEffxhVORL:-1037083546) ([raw file](https://github.com/cockroachdb/cockroach/blob/393c4a8e32f3d443441d0e68c0d903be3f813536/cli/cli_test.go#L540)):*
nope. the `select 3` was evaluated, and not the `select * from t.f`. Or at least, the result was from `select 3`. I'm poking at it and will file a bug.

---

*[cli/flags.go, line 25 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98AfGbm00FcuMx52Hs:-K98MD5JRYFx7Vubf3fA:982703159) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a5ef98ef4a86afdf7539e6b9adbb40d54e9d3c/cli/flags.go#L25)):*
you mean why did it show up? I'm using `security.RootUser` as the default value for `--user`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-:-K98MPBoArU8qRkqEI00:-360504021)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cli/flags.go, line 25 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98AfGbm00FcuMx52Hs:-K98Mti7iYeZBqDE1lUE:501287066) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a5ef98ef4a86afdf7539e6b9adbb40d54e9d3c/cli/flags.go#L25)):*
Oh, reviewable is fucking with me. This import should be below the non-cockroach imports though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Aside from the messy commit history, this LGTM. Maybe some of the commits could be squashed, though.

mberhault commented:
Review status: 16 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

*[cli/flags.go, line 25 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98AfGbm00FcuMx52Hs:-K98O-jWZdln9GvOWKug:845349431) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a5ef98ef4a86afdf7539e6b9adbb40d54e9d3c/cli/flags.go#L25)):*
are we putting a newline between `github.com/<non-cockroach>` and `github.com/cockroachdb`?  I don't see that being consistently applied.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[cli/flags.go, line 25 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98AfGbm00FcuMx52Hs:-K98OxdHpmvOtFhFyP9J:-1597396821) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a5ef98ef4a86afdf7539e6b9adbb40d54e9d3c/cli/flags.go#L25)):*
yeah, it's not consistent, but it seems to be the prevailing pattern when people consciously look at the import list

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 13 of 17 files reviewed at latest revision, 6 unresolved discussions.

---

*[cli/flags.go, line 25 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-K98AfGbm00FcuMx52Hs:-K98QLQx299_k7fgYV-2:-1057405798) ([raw file](https://github.com/cockroachdb/cockroach/blob/33a5ef98ef4a86afdf7539e6b9adbb40d54e9d3c/cli/flags.go#L25)):*
Done for the files I'm poking in this PR. It's unfortunate that we can't codify this in gofmt.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r16.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757)*
<!-- Sent from Reviewable.io -->

mberhault commented:
yup. I think all of them. even my original commit split of alphabetizing things got messy when rebasing.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3757#-:-K98OZAtaYxzSV0BVBL_:749097002)*
<!-- Sent from Reviewable.io -->
